image/tarexport/save_oci.go

```diff
@@ -0,0 +1,333 @@
+					return nil, err
+				}
+			}
+			continue
+		}
+		id, err = l.rs.Get(ref)
+		if err != nil {
+			return nil, err
+		}
+		if err := addAssoc(image.IDFromDigest(id), ref); err != nil {
+			return nil, err
+		}
+	}
+	return imgDescr, nil
+}
+
+func (s *ociSaveSession) save(outStream io.Writer) error {
+	s.diffIDsCache = make(map[layer.DiffID]*layerInfo)
+	s.savedImages = make(map[image.ID][]byte)
+	tempDir, err := ioutil.TempDir("", "oci-export-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	for id, info := range s.images {
+		for _, i := range info.refs {
+			ociRef, err := oci.NewReference(tempDir, i.Tag())
+			if err != nil {
+				return err
+			}
+			ociDest, err := ociRef.NewImageDestination(nil)
+			if err != nil {
+				return err
+			}
+			// TODO(runcom): handle foreign srcs like save.go
+			if err := s.saveImage(id, ociDest); err != nil {
+				return err
+			}
+		}
+		if len(info.refs) == 0 {
+			ociRef, err := oci.NewReference(tempDir, id.Digest().Hex())
+			if err != nil {
+				return err
+			}
+			ociDest, err := ociRef.NewImageDestination(nil)
+			if err != nil {
+				return err
+			}
+			// TODO(runcom): handle foreign srcs like save.go
+			if err := s.saveImage(id, ociDest); err != nil {
+				return err
+			}
+		}
+	}
+
+	fs, err := archive.Tar(tempDir, archive.Uncompressed)
+	if err != nil {
+		return err
+	}
+	defer fs.Close()
+
+	_, err = io.Copy(outStream, fs)
+	return err
+}
+
+func (s *ociSaveSession) saveImage(id image.ID, ociDest ctypes.ImageDestination) error {
+	if m, ok := s.savedImages[id]; ok {
+		// just add a new ref under refs/
+		if err := ociDest.PutManifest(m); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	img, err := s.is.Get(id)
+	if err != nil {
+		return err
+	}
+
+	if len(img.RootFS.DiffIDs) == 0 {
+		return fmt.Errorf("empty export - not implemented")
+	}
+
+	configJSON := img.RawJSON()
+	configInfo, err := ociDest.PutBlob(bytes.NewReader(configJSON), ctypes.BlobInfo{Size: -1})
+	if err != nil {
+		return err
+	}
+
+	// TODO(runcom): there should likely be a manifest builder (like docker/distribution)
+	m := imgspecv1.Manifest{
+		Versioned: imgspec.Versioned{
+			SchemaVersion: 2,
+			MediaType:     imgspecv1.MediaTypeImageManifest,
+		},
+		Config: imgspecv1.Descriptor{
+			MediaType: imgspecv1.MediaTypeImageConfig,
+			Digest:    configInfo.Digest,
+			Size:      configInfo.Size,
+		},
+	}
+
+	for i := range img.RootFS.DiffIDs {
+		rootFS := *img.RootFS
+		rootFS.DiffIDs = rootFS.DiffIDs[:i+1]
+
+		l, err := s.ls.Get(rootFS.ChainID())
+		if err != nil {
+			return err
+		}
+		defer layer.ReleaseAndLog(s.ls, l)
+
+		var (
+			digest string
+			size   int64
+		)
+		if i, ok := s.diffIDsCache[l.DiffID()]; ok {
+			digest = i.digest
+			size = i.size
+		} else {
+			arch, err := l.TarStream()
+			if err != nil {
+				return err
+			}
+			defer arch.Close()
+
+			// FIXME: anywhere I can get a gzipped layer (and digest) as found in remote registries?
```

Comment:
This aspect is mildly tricky. Just putting a comment here so I can take peak later.